### ISSUE TITLE
Issue: Unable to create resource

### ISSUE THREAD

**Issue by jeveleth** [sullivtr/terraform-provider-graphql#23]
Thanks for creating this module. I'm using it to configure an instance of Hasura Cloud (which uses a GraphQL API). Unfortunately, I'm encountering an issue while creating a simple record representing an image file with a string description.

```
2020/10/30 11:57:38 [INFO] backend/local: apply calling Apply
2020/10/30 11:57:38 [INFO] terraform: building graph: GraphTypeApply
2020/10/30 11:57:38 [DEBUG] ProviderTransformer: "graphql_mutation.basic_mutation (destroy)" (*terraform.NodeDestroyResourceInstance) needs provider["registry.terraform.io/sullivtr/graphql"]
2020/10/30 11:57:38 [DEBUG] ProviderTransformer: "graphql_mutation.basic_mutation (expand)" (*terraform.nodeExpandApplyableResource) needs provider["registry.terraform.io/sullivtr/graphql"]
2020/10/30 11:57:38 [DEBUG] ProviderTransformer: "graphql_mutation.basic_mutation" (*terraform.NodeApplyableResourceInstance) needs provider["registry.terraform.io/sullivtr/graphql"]
2020/10/30 11:57:38 [DEBUG] ReferenceTransformer: "graphql_mutation.basic_mutation (expand)" references: []
2020/10/30 11:57:38 [DEBUG] ReferenceTransformer: "graphql_mutation.basic_mutation" references: []
2020/10/30 11:57:38 [DEBUG] ReferenceTransformer: "provider[\"registry.terraform.io/sullivtr/graphql\"]" references: []
2020/10/30 11:57:38 [DEBUG] Starting graph walk: walkApply
2020-10-30T11:57:38.088-0700 [INFO]  plugin: configuring client automatic mTLS
2020-10-30T11:57:38.112-0700 [DEBUG] plugin: starting plugin: path=.terraform/plugins/registry.terraform.io/sullivtr/graphql/1.4.5/darwin_amd64/terraform-provider-graphql_1.4.5 args=[.terraform/plugins/registry.terraform.io/sullivtr/graphql/1.4.5/darwin_amd64/terraform-provider-graphql_1.4.5]
2020-10-30T11:57:38.115-0700 [DEBUG] plugin: plugin started: path=.terraform/plugins/registry.terraform.io/sullivtr/graphql/1.4.5/darwin_amd64/terraform-provider-graphql_1.4.5 pid=11785
2020-10-30T11:57:38.115-0700 [DEBUG] plugin: waiting for RPC address: path=.terraform/plugins/registry.terraform.io/sullivtr/graphql/1.4.5/darwin_amd64/terraform-provider-graphql_1.4.5
2020-10-30T11:57:38.126-0700 [INFO]  plugin.terraform-provider-graphql_1.4.5: configuring server automatic mTLS: timestamp=2020-10-30T11:57:38.126-0700
2020-10-30T11:57:38.151-0700 [DEBUG] plugin: using plugin: version=5
2020-10-30T11:57:38.151-0700 [DEBUG] plugin.terraform-provider-graphql_1.4.5: plugin address: address=/var/folders/ly/k647m4vs1tn_2n773hvts4yr0000gn/T/plugin114109308 network=unix timestamp=2020-10-30T11:57:38.151-0700
2020-10-30T11:57:38.200-0700 [WARN]  plugin.stdio: received EOF, stopping recv loop: err="rpc error: code = Unimplemented desc = unknown service plugin.GRPCStdio"
2020/10/30 11:57:38 [DEBUG] EvalApply: ProviderMeta config value set
2020/10/30 11:57:38 [DEBUG] graphql_mutation.basic_mutation: applying the planned Delete change
graphql_mutation.basic_mutation: Destroying... [id=181708519]
graphql_mutation.basic_mutation: Destruction complete after 1s
graphql_mutation.basic_mutation: Creating...
2020/10/30 11:57:38 [DEBUG] EvalApply: ProviderMeta config value set
2020/10/30 11:57:38 [DEBUG] graphql_mutation.basic_mutation: applying the planned Create change
2020-10-30T11:57:38.757-0700 [DEBUG] plugin.terraform-provider-graphql_1.4.5: 2020/10/30 11:57:38 [DEBUG] setting computed for "computed_read_operation_variables" from ComputedKeys
2020-10-30T11:57:38.757-0700 [DEBUG] plugin.terraform-provider-graphql_1.4.5: 2020/10/30 11:57:38 [DEBUG] setting computed for "computed_update_operation_variables" from ComputedKeys
2020-10-30T11:57:38.757-0700 [DEBUG] plugin.terraform-provider-graphql_1.4.5: 2020/10/30 11:57:38 [DEBUG] setting computed for "computed_delete_operation_variables" from ComputedKeys
2020-10-30T11:57:38.944-0700 [DEBUG] plugin.terraform-provider-graphql_1.4.5: 2020/10/30 11:57:38 [ERROR] Unable to compute mutation variable keys: mutation_key not found
graphql_mutation.basic_mutation: Creation complete after 0s [id=3498629838]
2020-10-30T11:57:38.957-0700 [DEBUG] plugin: plugin process exited: path=.terraform/plugins/registry.terraform.io/sullivtr/graphql/1.4.5/darwin_amd64/terraform-provider-graphql_1.4.5 pid=11785
2020-10-30T11:57:38.957-0700 [DEBUG] plugin: plugin exited
```

My configuration is below:

```
provider "graphql" {
  url = "REDACTED"
  headers = {
    "x-hasura-admin-secret" = "REDACTED"
    "Content-Type"          = "application/json"
  }
}

resource "graphql_mutation" "basic_mutation" {
  mutation_variables = {
    "descripton" = "fun times"
  }

  compute_from_create = false

  compute_mutation_keys = {
    "id" = "image.id"
  }

  create_mutation = file("./queries/createMutation.gql")
  update_mutation = file("./queries/updateMutation.gql")
  delete_mutation = file("./queries/deleteMutation.gql")
  read_query      = file("./queries/readQuery.gql")
}
```

The relevant files are: 

createMutation.gql file
```
mutation insertImage($description: String!) {
  insert_image(objects: {description: $description}) {
    returning {
      id
    }
  }
}
```

**Comment by miguelaferreira** [comment]
I'm having a similar issue trying to create github organizations. The resource is not created but the provider does not fail the terraform run.

**Comment by sullivtr** [comment]
Hello. Thank you for reporting this issue. According to the logs, the provider is failing when attempting to compute keys based on what you have provided inside of 
```
compute_mutation_keys = {
    "id" = "image.id"
}
```

Does your api respond with an object that looks like this: 
```
{
  data: {
    image: {
         id: "someIdValue"
    }
  }
}
```

The reason for the error: `Unable to compute mutation variable keys: mutation_key not found
graphql_mutation.basic_mutation: Creation complete after 0s [id=3498629838]` is that it was unable to find data.image.id in the response object. 

However, it appears that you also have "compute_from_create" set to false, which mean that it will try to traverse the response object based on the read query, and not the create mutation's response. If you set "compute_from_create" to true, does the provider successfully find the data.image.id value?

**Comment by sullivtr** [comment]
I should point out that another reason that the provider will fail to compute mutation keys is if the create mutation did not successfully create the resource you intended (or the api itself does not respond with what it created in its response body). Can you also confirm that the Hasura Cloud api call is happening, and that the resource is being created?

**Comment by sullivtr** [comment]
@miguelaferreira, with respect to error responses from the API. I may need to add better error handling in the query_executior itself with respect to api responses. I am curious if you know which error the github api is returning when the create call is initiated? Posting the mutation you are using may also help me solve your issue as well.

**Comment by jeveleth** [comment]
@sullivtr Thanks for the response. The response body from Hasura Cloud is actually:
```
{
  "data": {
    "image": [
      {
        "id": "someIdValue"
      }
    ]
  }
}
```

I've set `compute_from_create = true` and adjusted the mutation keys:
```
  compute_mutation_keys = {
    "id" = "image.[].id"
  }
```
The resource is not created, though I am able to successfully create the same resource via a curl call to the Hasura Cloud API.

**Comment by sullivtr** [comment]
Okay. This is helpful. Unfortunately, one of the limitations of the provider that the key computation cannot traverse arrays in the response object (this is noted in the documentation) (Since it wouldnt know which item to grab the key from in the list). If you add a feature request (PRs are welcome also) I can look into making this work, but its a bit tricky since the provider has to implicitly traverse the response object. However, I dont think it is impossible.

**Comment by jeveleth** [comment]
Ah. Got it. Thanks! I can open a feature request and/or PR when I get some time.

**Comment by sullivtr** [comment]
@jeveleth So I spent some time looking into the hasura cloud GQL api. In your case, it may not be the most terraformy thing to use this provider to orchestrate your GQL calls for Hasura Cloud (at least not for creating the images). That is, you may not actually want terraform to be _the source of truth_ for your hasura cloud images. This is a situation where I would write a custom terraform provider specifically for Hasura Cloud. There is one here: https://registry.terraform.io/providers/hasuracommunity/hasura/latest, but that seems to only have a resource for creating tenants (which, quite honestly, is the only thing that would make sense to manage via terraform)

**Comment by jeveleth** [comment]
Thanks! Sorry, I was using this merely as a proof of concept in an effort
to configure the Hasura Cloud instance itself via TF.

**Comment by sullivtr** [comment]
No need to apologize. Thanks for submitting the issue. Its got me thinking about how I can improve things potentially. Im mostly just bummed out that I was not able to unblock you at the moment.

**Comment by sullivtr** [comment]
@jeveleth @miguelaferreira I am reopening this issue now. I have started working on a solution to allow computations from nested objects. I have a working example already, but will need to clean things up/do more testing before cutting a PR. Stay tuned!

**Comment by sullivtr** [comment]
Here is my proposed solution, so feel free to give feedback: 

You will be able to locate keys this way: 
```hcl
compute_mutation_keys = {
   "some_nested_id" = "data.items[3].id
}
```
As well as multi nested: 
```hcl
compute_mutation_keys = {
   "some_nested_id" = "data.items[3].moreItems[2].id
}
```

**Comment by sullivtr** [comment]
Related PR: https://github.com/sullivtr/terraform-provider-graphql/pull/29

**Comment by ajbouh** [comment]
Is there a reason that the terraform operation still completes successfully after we see `Unable to compute mutation variable keys: mutation_key not found` in the logs?

It seems like this should be a fatal error, but perhaps I'm missing something?

**Comment by sullivtr** [comment]
@ajbouh Thank you for pointing this out. It appears to have been an oversight with logging. I'm going to file a separate issue related to this bug, and get it fixed when I have time (PRs are also welcome). This line of code here just needs to return an error diagnositic: https://github.com/sullivtr/terraform-provider-graphql/blob/master/graphql/resource_graphql_mutation.go#L200

**Comment by sullivtr** [comment]
> Is there a reason that the terraform operation still completes successfully after we see `Unable to compute mutation variable keys: mutation_key not found` in the logs?
> 
> It seems like this should be a fatal error, but perhaps I'm missing something?

@ajbouh Thank you for pointing this out. It appears to have been an oversight with logging. I'm going to file a separate issue related to this bug, and get it fixed when I have time (PRs are also welcome). This line of code here just needs to return an error diagnositic: https://github.com/sullivtr/terraform-provider-graphql/blob/master/graphql/resource_graphql_mutation.go#L200

**Comment by ajbouh** [comment]
I'm seeing this error and struggling to identify which key is missing. Perhaps the failure message could provide enough info to debug what's going on.

Logging the response at DEBUG level seems like a good thing along these lines.

Also seems like *any* error in the graphql response should be fatal, what do you think?
